src/sentry/migrations/0093_cleanup_unreferenced_event_files.py

```diff
@@ -0,0 +1,64 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.28 on 2020-07-28 16:24
+from __future__ import unicode_literals
+
+from django.db import migrations
+from django.db.models import Min
+
+from sentry.utils.query import RangeQuerySetWrapperWithProgressBar
+
+def cleanup_event_attachment_files(apps, schema_editor):
+    """
+    Previously, cleanup task code did a SQL bulk delete on EventAttachment
+    leaving orphaned File and FileBlob objects. These orphaned files now need to
+    be purged as they are still consuming space.
+    """
+    EventAttachment = apps.get_model("sentry", "EventAttachment")
+    File = apps.get_model("sentry", "File")
+
+    # Find the oldest live attachment as we only want to purge old files.
+    oldest_attachment = EventAttachment.objects.all().aggregate(Min("date_added"))
+
+    # File types used in event attachments.
+    attachment_types = [
+        "event.applecrashreport",
+        "event.attachment",
+        "event.payload",
+        "event.minidump",
+        "unreal.context",
+        "unreal.logs",
+    ]
+    file_query = (File.objects
+        .filter(timestamp__lt=oldest_attachment['date_added__min'])
+        .filter(type__in=attachment_types))
+
+    for f in RangeQuerySetWrapperWithProgressBar(file_query):
+        # Double check that the file is not referenced.
+        if not EventAttachment.objects.filter(file=f).exists():
+            f.delete()
```

Comment:
Do we also need to delete the related `FileBlob` rows here? And if so, do we expect that this should also delete the file itself from storage?

Reply:
Short answer, no.
FileBlobs are a many-to-many relationship with FileBlobIndex and File.
The cleanup to remove dangling Fileblobs is done [here](https://github.com/getsentry/sentry/blob/master/src/sentry/runner/commands/cleanup.py#L328-L352).

Reply:
The File model is a curious beast. It uses relationship cascades to delete the blob models [which overload delete() to clean up storage](https://github.com/getsentry/sentry/blob/master/src/sentry/models/file.py#L263-L286).

Similar style deletions for [releases use the .delete() method on ReleaseFile](https://github.com/getsentry/sentry/blob/e0c247c17c59a72f2241d972933dfd3203ecf0be/src/sentry/deletions/defaults/organization.py#L46) and rely on django cascading deletes into File and FileBlob which is what I used as prior art for this.

Reply:
I don't think we can rely on the custom `delete` from the `FileBlob` in migrations, these models are just 'fake' versions of the models. So I don't think that will actually happen. It also would not cascade across a M2M relationship, and afaik the FK relationship is deprecated.

Anyway, based on what @JTCunning said the file blobs will be cleaned up via a separate process, so we should be good here.

Reply:
I had forgotten that django uses fake models in migrations, the blob cleanup will catch the 2nd tier of orphans :smile: